File: solutions/javascript/class-design/implementTrie.js
class Trie {
    constructor() {
        this.trie = {};
        this.endSymbol = '*';
    }

    insert(word) {
        let curPosition = this.trie;

        for (const char of word) {
            const charInCurPosition = curPosition.hasOwnProperty(char);

            if (!charInCurPosition) curPosition[char] = {};
            curPosition = curPosition[char];
        }

        curPosition[this.endSymbol] = true;
    }

    search(word) {
        let curPosition = this.trie;

        for (const char of word) {
            const charInCurPosition = curPosition.hasOwnProperty(char);

            if (!charInCurPosition) return false;
            curPosition = curPosition[char];
        }

        const endSymbolExists = curPosition.hasOwnProperty(this.endSymbol);

        return endSymbolExists;
    }

    startsWith(prefix) {
        let curPosition = this.trie;

        for (const char of prefix) {
            const charInCurPosition = curPosition.hasOwnProperty(char);

            if (!charInCurPosition) return false;
            curPosition = curPosition[char];
        }

        return true;
    }
}

File: solutions/python/class-design/implement_trie.py
class Trie:
    def __init__(self):
        self.trie = {}
        self.end_symbol = "*"

    def insert(self, word: str) -> None:
        cur_position = self.trie

        for char in word:
            if char not in cur_position:
                cur_position[char] = {}
            cur_position = cur_position[char]

        cur_position[self.end_symbol] = True

    def search(self, word: str) -> bool:
        cur_position = self.trie

        for char in word:
            if char not in cur_position:
                return False
            cur_position = cur_position[char]

        return self.end_symbol in cur_position

    def startsWith(self, prefix: str) -> bool:
        cur_position = self.trie

        for char in prefix:
            if char not in cur_position:
                return False
            cur_position = cur_position[char]

        return True